habitat_baselines/il/data/eqa_cnn_pretrain_data.py

```diff
@@ -0,0 +1,227 @@
+            for ep in self.episodes:
+                if ep.scene_id not in self.scene_ids:
+                    self.scene_ids.append(ep.scene_id)
+                    self.scene_episode_dict[ep.scene_id] = [ep.episode_id]
+                else:
+                    self.scene_episode_dict[ep.scene_id].append(ep.episode_id)
+
+            self.train_lmdb = lmdb.open(
+                self.dataset_path.format(split="train"),
+                map_size=int(1e11),
+                writemap=True,
+            )
+
+            self.val_lmdb = lmdb.open(
+                self.dataset_path.format(split="val"),
+                map_size=int(0.5e11),
+                writemap=True,
+            )
+
+            self.train_count = -1
+            self.val_count = -1
+
+            for scene in tqdm(list(self.scene_episode_dict.keys())):
+
+                self.load_scene(scene)
+
+                for ep_id in tqdm(self.scene_episode_dict[scene]):
+
+                    episode = next(
+                        ep for ep in self.episodes if ep.episode_id == ep_id
+                    )
+                    try:
+                        # TODO: Consider alternative for shortest_paths
+                        pos_queue = episode.shortest_paths[0]
+                    except AttributeError as e:
+                        logger.error(e)
+
+                    random_pos = random.sample(pos_queue, 9)
+                    self.save_frames(random_pos)
+
+            logger.info("EQA-CNN-PRETRAIN database ready!")
+
+            if self.mode == "train":
+                lmdb_env = self.train_lmdb
+            elif self.mode == "val":
+                lmdb_env = self.val_lmdb
+
+            self.env.close()
+
+        else:
+            lmdb_env = lmdb.open(
+                self.dataset_path.format(split=self.mode),
+                readonly=True,
+                lock=False,
+            )
+
+        self.lmdb_txn = lmdb_env.begin()
+        self.lmdb_cursor = self.lmdb_txn.cursor()
+
+    def save_frames(self, pos_queue: List[ShortestPathPoint]) -> None:
+        r"""Writes rgb, seg, depth frames to LMDB.
+        """
+
+        for pos in pos_queue:
+
+            observation = self.env.sim.get_observations_at(
+                pos.position, pos.rotation
+            )
+
+            depth = observation["depth"]
+            rgb = observation["rgb"]
+
+            scene = self.env.sim.semantic_annotations()
+            instance_id_to_label_id = {
+                int(obj.id.split("_")[-1]): obj.category.index()
+                for obj in scene.objects
+            }
+            self.mapping = np.array(
+                [
+                    instance_id_to_label_id[i]
+                    for i in range(len(instance_id_to_label_id))
+                ]
+            )
+            seg = np.take(self.mapping, observation["semantic"])
+            seg[seg == -1] = 0
+            seg = seg.astype("uint8")
+
+            if random.random() < 0.8:
```

Comment:
Didn't notice this before, but this is absolutely cannot be how train/val is split -- it would mean you are pre-training the CNN on the scenes we consider as held-out for evaluating the agent.  Please use the train-val split by scene that already exists.

Reply:
Thanks for pointing this out @erikwijmans . My inital thought was that the feature extractor could be agnostic to the train/test slipt for other tasks because feature extraction is more of a generic task. I'll make the changes. Thanks

Reply:
Yeah, I see where you are coming from, but we can no longer call the val scenes held-out of the encoder was trained on them.

Reply:
Got it, thanks!